### PR TITLE
 [FLINK-21072] Refactor the SnapshotStrategy hierarchy 

### PR DESCRIPTION
**Pull request by dawidwys**
## What is the purpose of the change

1. The refactoring simplifies the snapshot strategy hierarchy a bit.

2. It extracts the synchronous/asynchronous execution out of the SnapshotStrategy to a separate SnapshotStrategyRunner. Prior to those changes the logic was duplicated in most of the strategies.

3. It introduces the concept of SnapshotResources for the synchronous part which can be later on used for introducing a common SavepointSnapshotResources.

## Verifying this change

This change is already covered by existing state backend tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SnapshotStrategyRunner.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    @Nonnull private final SnapshotStrategy<T, SR> snapshotStrategy;
+    @Nonnull private final CloseableRegistry cancelStreamRegistry;
+
+    @Nonnull private final boolean isSynchronous;
+
+    public SnapshotStrategyRunner(
+            @Nonnull String description,
+            @Nonnull SnapshotStrategy<T, SR> snapshotStrategy,
+            @Nonnull CloseableRegistry cancelStreamRegistry,
+            boolean isSynchronous) {
```
Comment (aljoscha) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/SnapshotStrategyRunner.java:62]: nit: I was already confused about what `true` or `false` meant when I saw callsites. Maybe some enum with `SYNCHRONOUS` `ASYNCHRONOUS` would help.

**Review by aljoscha — CHANGES_REQUESTED** [review]
The changes look very good overall!

I had some nitpicks and suggestions about naming.

**Review by aljoscha — APPROVED** [review]
Changes look good! Ready to merge once CI is ✅

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SnapshotStrategyRunner.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+                snapshotStrategy.asyncSnapshot(
+                        snapshotResources,
+                        checkpointId,
+                        timestamp,
+                        streamFactory,
+                        checkpointOptions);
+
+        switch (executionType) {
+            case SYNCHRONOUS:
+                return DoneFuture.of(asyncSnapshot.get(cancelStreamRegistry));
```
Comment (aljoscha) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/SnapshotStrategyRunner.java:93]: Is this using the same registry as before? (Say in the `HeapSnapshotStrategy`)

**Review by aljoscha — APPROVED** [review]
I believe the changes look good now! I had one inline question, just for my understanding.

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/SnapshotStrategyRunner.java (hunk 2 above)
Comment (dawidwys) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/SnapshotStrategyRunner.java:93]: Good catch. I think this is actually a bug, as it is using a different one. With this registry the main checkpointing thread(from which the `snapshot` method is executed) might not get cancelled before shutting down the io streams.

I will rewrite the code to always go through the `AsyncSnapshotTask`

**Pull request closed without merging**